php-checks/src/main/java/org/sonar/php/checks/OverwrittenArrayElementCheck.java

```diff
@@ -0,0 +1,170 @@
+  private Map<String, Map<String, AssignmentExpressionTree>> writtenAndUnread = new HashMap<>();
+
+  @Override
+  public void visitNode(Tree tree) {
+    namesToSymbols.clear();
+    writtenAndUnread.clear();
+
+    List<StatementTree> statementTrees;
+    if (tree.is(Tree.Kind.BLOCK)) {
+      statementTrees = ((BlockTree) tree).statements();
+    } else {
+      statementTrees = ((ScriptTree) tree).statements();
+    }
+
+    for (StatementTree statementTree : statementTrees) {
+      if (!isArrayKeyAssignmentStatement(statementTree)) {
+        removeReadArrayKeys(statementTree);
+        continue;
+      }
+
+      AssignmentExpressionTree assignmentExpressionTree = (AssignmentExpressionTree) ((ExpressionStatementTree) statementTree).expression();
+      ArrayAccessTree arrayAccessTree = (ArrayAccessTree) assignmentExpressionTree.variable();
+      String key = ((LiteralTree) arrayAccessTree.offset()).value();
+      String variableName = ((VariableIdentifierTree) arrayAccessTree.object()).text();
+      Symbol variableSymbol = context().symbolTable().getSymbol(arrayAccessTree.object());
+
+      checkArrayKeyWrite(statementTree, assignmentExpressionTree, key, variableName, variableSymbol);
+
+      updateWrittenAndUnread(assignmentExpressionTree, key, variableName, variableSymbol);
+    }
+
+    super.visitNode(tree);
+  }
+
+  private void checkArrayKeyWrite(StatementTree statementTree,
+    AssignmentExpressionTree assignmentExpressionTree,
+    String key,
+    String variableName,
+    Symbol variableSymbol) {
+    if (writtenAndUnread.containsKey(variableName) &&
+      writtenAndUnread.get(variableName).containsKey(key) &&
+      !symbolWasUsedInTree(variableSymbol, assignmentExpressionTree.value())) {
+      AssignmentExpressionTree firstAssignmentTree = writtenAndUnread.get(variableName).get(key);
+      context().newIssue(this, statementTree, MESSAGE).secondary(firstAssignmentTree, MESSAGE_SECONDARY);
+    }
+  }
+
+  private void updateWrittenAndUnread(AssignmentExpressionTree statementTree, String key, String variableName, Symbol variableSymbol) {
+    Map<String, AssignmentExpressionTree> keysToAssignmentMap;
+    if (writtenAndUnread.containsKey(variableName)) {
+      keysToAssignmentMap = writtenAndUnread.get(variableName);
+    } else {
+      keysToAssignmentMap = new HashMap<>();
+    }
+    keysToAssignmentMap.put(key, statementTree);
+    namesToSymbols.put(variableName, variableSymbol);
+    writtenAndUnread.put(variableName, keysToAssignmentMap);
+  }
+
+  private void removeReadArrayKeys(StatementTree statementTree) {
+    Map<String, Map<String, AssignmentExpressionTree>> cleanedWrittenAndUnread = new HashMap<>();
+    for (Map.Entry<String, Map<String, AssignmentExpressionTree>> entry : writtenAndUnread.entrySet()) {
+      if (!symbolWasUsedInTree(namesToSymbols.get(entry.getKey()), statementTree)) {
+        cleanedWrittenAndUnread.put(entry.getKey(), entry.getValue());
+      }
+    }
```

Comment:
Why do we need those lines? Unit tests are still green if I remove them.

Reply:
Strangely there wasn't a test case that demonstrated why this is needed. Added one. (And also rewrote it by using Streams)